### PR TITLE
Use gRPC and protobuf in bessctl

### PR DESCRIPTION
**Pull request by changlan**
- Use gRPC and protobuf in libbess-python and bessctl
- Obsolete master.cc and use bessctl.cc instead

### CODE REVIEW

**Comment by sangjinhan** [comment]
Yay, super excited!!

File: libbess-python/bess.py [hunk 1]
```diff
@@ -32,188 +39,241 @@ class APIError(Exception):
...
+
+        all_classes = inspect.getmembers(module_msg, lambda c: inspect.isclass(c))
+        arg_classes = dict(all_classes)
+        arg_classes['EmptyArg'] = bess_msg.EmptyArg
+
+        arg_msg = proto_conv.dict_to_protobuf(arg, arg_classes[arg_type])
+        request.arg.Pack(arg_msg)
+
+        response = self._request(self.stub.ModuleCommand, request)
+        if response.HasField('other'):
```
Comment (sangjinhan) [inline libbess-python/bess.py]: Is this temporary, or will we have to update here whenever we introduce new response types?

File: libbess-python/bess_test.py [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+        client.connect()
+
+        response = client.list_modules()
+        self.assertEqual(0, response.error.err)
+
+    def test_create_port(self):
+        client = bess.BESS()
+        client.connect()
+
+        response = client.create_port('pcap_port', 'p0', {'dev': 'rnd'})
```
Comment (sangjinhan) [inline libbess-python/bess_test.py]: pcap_port -> PCAPPort (the C++ class name is exported). Other port names below should updated as well

File: libbess-python/bess_test.py [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+        self.assertEqual('p0', response.name)
+
+        response = client.create_port('vport', 'p0', {
+            'ifname': 'veth0',
+            'container_pid': 23124,
+            'rxq_cpus': [1,2,3],
+            'tx_tci': 123,
+            'tx_outer_tci':123,
+            'loopback': False,
+            'ip_addrs': ['1.1.1.1', '2.2.2.2']
```
Comment (sangjinhan) [inline libbess-python/bess_test.py]: IP addresses that are not palindrome would be better, to test if endianness works correctly

File: protobuf/port_msg.proto [hunk 4]
```diff
@@ -7,8 +7,8 @@ message PCAPPortArg {
 }
 
 message PMDPortArg {
-  bool loopback = 1;
-  uint32 port_id = 2;
+  uint64 loopback = 1;
```
Comment (sangjinhan) [inline protobuf/port_msg.proto]: Why not bool?

**Comment by sangjinhan** [comment]
Some issues with bessctl:
- Ctrl + D does nothing (supposed to terminate the interactive shell)
- `quit` command incurs the following error message:
```
E1117 23:27:45.520209496  177461 subchannel_index.c:70]      assertion failed: c != NULL
Aborted (core dumped)
```

File: core/bessctl.cc [hunk 5]
```diff
@@ -189,6 +190,7 @@ static ::Port* create_port(const std::string& name, const PortBuilder& driver,
   p->queue_size[PACKET_DIR_INC] = size_inc_q;
   p->queue_size[PACKET_DIR_OUT] = size_out_q;
 
+  ctx.SetNonWorker();
```
Comment (changlan) [inline core/bessctl.cc]: Agreed. I can't find any knobs in gRPC for that though. @barath Did you have any experience?

Comment (barath) [inline core/bessctl.cc]: gRPC has way way fewer knobs than stubby -- I don't know of anything here.

File: libbess-python/bess.py (hunk 1 above)
Comment (changlan) [inline libbess-python/bess.py]: Fixed. Now it can lookup response types dynamically.

**Comment by changlan** [comment]
All addressed except the ctx.SetNonWorker() one.

**Comment by sangjinhan** [comment]
Thanks! Could you take a look at the following issues? Sorry for this late response.
- `daemon start` does not immediately refresh the connection status
```
<disconnected> $ daemon start
<disconnected> $
localhost:10514 $
```
- `daemon stop` has the same issue, in the other way around
- The following commands do not work (due to the same problem?)
  - `show pipeline`
  - `monitor pipeline`
  - `show module`
  - `show module [MODULE]`
  - `show status`
- The following commands do not work (due to another problem?)
  - `show tc`
- Auto completion for driver names in `show driver DRIVER...` does not work
- `track` command seems not ported
- `monitor port` breaks due to the lack of operator overloading

I will merge #140 first in the meantime. It may introduce minor merge conflicts with this PR, then could you please rebase with the latest c++ upstream?

**Comment by changlan** [comment]
Gotcha. Patches that fix those issues have been committed.

**Comment by sangjinhan** [comment]
Merged, thanks!
img

**Pull request merged**